core/src/main/java/com/linecorp/armeria/common/AbstractHttpMessageBuilder.java

```diff
@@ -0,0 +1,130 @@
+import com.linecorp.armeria.internal.common.JacksonUtil;
+
+abstract class AbstractHttpMessageBuilder {
+
+    @Nullable
+    private HttpData content;
+
+    @Nullable
+    private Publisher<? extends HttpData> publisher;
+
+    @Nullable
+    protected HttpData getContent() {
+        return content;
+    }
+
+    @Nullable
+    protected Publisher<? extends HttpData> getPublisher() {
+        return publisher;
+    }
+
+    /**
+     * Sets the content for this response.
+     */
+    public AbstractHttpMessageBuilder content(HttpData content) {
+        requireNonNull(content, "content");
+        this.content = content;
+        return this;
+    }
+
+    /**
+     * Sets the {@link Publisher} for this response.
+     */
+    public AbstractHttpMessageBuilder content(Publisher<? extends HttpData> content) {
+        requireNonNull(content, "publisher");
+        checkState(this.content == null, "content has been set already");
+        publisher = content;
+        return this;
+    }
+
+    /**
+     * Sets the content as UTF_8 for this response.
+     */
+    public abstract AbstractHttpMessageBuilder content(String content);
+
+    /**
+     * Sets the content for this response.
+     */
+    public abstract AbstractHttpMessageBuilder content(MediaType contentType, CharSequence content);
+
+    /**
+     * Sets the content for this response.
+     */
+    public abstract AbstractHttpMessageBuilder content(MediaType contentType, String content);
+
+    /**
+     * Sets the content as UTF_8 for this response. The {@code content} is formatted by
+     * {@link String#format(Locale, String, Object...)} with {@linkplain Locale#ENGLISH English locale}.
+     */
+    @FormatMethod
+    public abstract AbstractHttpMessageBuilder content(@FormatString String format, Object... content);
+
+    /**
+     * Sets the content for this response. The {@code content} is formatted by
+     * {@link String#format(Locale, String, Object...)} with {@linkplain Locale#ENGLISH English locale}.
+     */
+    @FormatMethod
+    public abstract AbstractHttpMessageBuilder content(MediaType contentType, @FormatString String format,
+                                                       Object... content);
+
+    /**
+     * Sets the content for this response. The {@code content} will be wrapped using
+     * {@link HttpData#wrap(byte[])}, so any changes made to {@code content} will be reflected in the response.
+     */
+    public abstract AbstractHttpMessageBuilder content(MediaType contentType, byte[] content);
+
+    /**
+     * Sets the content for this response.
+     */
+    public abstract AbstractHttpMessageBuilder content(MediaType contentType, HttpData content);
+
+    /**
+     * Sets the {@link Publisher} for this response.
+     */
+    public abstract AbstractHttpMessageBuilder content(MediaType contentType,
+                                                       Publisher<? extends HttpData> content);
```

Comment:
I think we can remove `abstract` from some methods and delegate to other methods that could simplify the implementation of subclasses.
For example:
```java
public AbstractHttpMessageBuilder content(MediaType contentType, byte[] content) {
    return content(contentType, HttpData.wrap(content));
}
```